### PR TITLE
[wip] Add shim layer to abstract dependencies on mbed OS

### PR DESCRIPTION
**Pull request by salkinium**
This allows `utest` to be used outside of mbed OS.

There are two scheduler implementations in the shim:
- using MINAR for mbed OS (default),
- using `us_ticker` for mbed OS (and perhaps mbedmicro?)

If yotta is detected, MINAR will be used, else `us_ticker` is used.

To use `utest` outside of mbed OS you also need to provide a critical section implementation using these two macros:
- `UTEST_ENTER_CRITICAL_SECTION`, and
- `UTEST_LEAVE_CRITICAL_SECTION`.

The usage of the macros inside the Harness is compatible with RAII, so it is possible to use a RAII class for its purpose:

```
#define UTEST_ENTER_CRITICAL_SECTION YourRAIIClass lock
```

Note: This code has not been tested with mbedmicro.

cc @0xc0170 @PrzemekWirkus

### CODE REVIEW

**Comment by PrzemekWirkus** [comment]
> If yotta is detected, MINAR will be used, else us_ticker is used.

Maybe assuming that using ```yotta``` will automatically enable ```MINAR``` is not the best idea.
On the other hand we should be backward compatible with previous implementations.

**Comment by niklas-arm** [comment]
> Maybe assuming that using yotta will automatically enable MINAR is not the best idea.

It actually checks if `YOTTA_MINAR_VERSION_STRING` is defined, which it is only, if you included the module in yotta.

**Comment by 0xc0170** [comment]
How to introduce custom scheduler into the utest code base? would utest_config header file be a good answer? Or to tweak a build configuration to get those macros in? If a config header file, how to make it backward compatible within this PR?

I went with utest config for a quick test of this new API, I created utest_config.h:

```
// just a test, don't take this seriously
#define UTEST_ENTER_CRITICAL_SECTION __disable_irq()
#define UTEST_LEAVE_CRITICAL_SECTION __enable_irq()

// custom scheduler
#define YOTTA_CFG_UTEST_USE_CUSTOM_SCHEDULER 1
```

I had to add ``#include utest_config.h`` to shim.h (= using external build system, which does not allow me to add macros), implemented my own custom scheduling via provided API, and was able to run it.

**Comment by niklas-arm** [comment]
Hm, ok, so different idea without using macros: If you want to port utest to another platform, you need to implement these three functions:

- `void utest_enter_critical_section(void)`
- `void utest_leave_critical_section(void)`
- `utest_v1_scheduler_t utest_get_scheduler(void)`

Similar to how it is done now, there would be default implementations for when use of Minar and Core Util modules in yotta have been detected, otherwise the functions will have to be implemented and linked externally.

File: utest/shim.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#   ifdef YOTTA_MINAR_VERSION_STRING
+#       define UTEST_MINAR_AVAILABLE 1
+#   else
+#       define UTEST_MINAR_AVAILABLE 0
+#   endif
+#   ifndef UTEST_SHIM_SCHEDULER_USE_MINAR
+#       define UTEST_SHIM_SCHEDULER_USE_MINAR UTEST_MINAR_AVAILABLE
+#   endif
+#   ifndef UTEST_SHIM_SCHEDULER_USE_US_TICKER
+#       define UTEST_SHIM_SCHEDULER_USE_US_TICKER (!UTEST_MINAR_AVAILABLE)
```
Comment (0xc0170) [inline utest/shim.h]: This effectively enables us ticker if a user does not supply any of minar version string or use minar. Shouldn't we leave it (no macros defined for utest), and a user needs to to implement the required API?

Comment (niklas-arm) [inline utest/shim.h]: Whooops

File: source/shim.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+{
+    return utest_v1_scheduler;
+}
+}
+
+#elif UTEST_SHIM_SCHEDULER_USE_US_TICKER
+// only one callback is active at any given time
+static volatile utest_v1_harness_callback_t minimal_callback;
+static volatile utest_v1_harness_callback_t ticker_callback;
+static const ticker_data_t *ticker_data;
```
Comment (0xc0170) [inline source/shim.cpp]: This should pull in the us ticker header file

Comment (niklas-arm) [inline source/shim.cpp]: Something like this? `#include "us_ticker.h"` is the correct path for mbed?
```c
#ifdef YOTTA_MBED_HAL_VERSION_STRING
#   include "mbed-hal/us_ticker.h"
#else
#   include "us_ticker.h"
#endif
```

Comment (0xc0170) [inline source/shim.cpp]: Yes that should do it

File: utest/shim.h [hunk 3]
```diff
@@ -59,7 +59,7 @@
 #       define UTEST_SHIM_SCHEDULER_USE_MINAR UTEST_MINAR_AVAILABLE
 #   endif
 #   ifndef UTEST_SHIM_SCHEDULER_USE_US_TICKER
-#       define UTEST_SHIM_SCHEDULER_USE_US_TICKER (!UTEST_MINAR_AVAILABLE)
+#       define UTEST_SHIM_SCHEDULER_USE_US_TICKER 0
```
Comment (0xc0170) [inline utest/shim.h]: would it make sense to enable us ticker by default for mbed SDK (__MBED__=1) ?
```
#       if defined(__MBED__) && (__MBED__ == 1)
#           define UTEST_SHIM_SCHEDULER_USE_US_TICKER 1
#       else
#           define UTEST_SHIM_SCHEDULER_USE_US_TICKER 0
#       endif
```

Comment (niklas-arm) [inline utest/shim.h]: I added you as a collaborator, so you can make the changes faster, as I cannot test this right now?

Comment (niklas-arm) [inline utest/shim.h]: And to the original question: Yes, I think so.

**Comment by 0xc0170** [comment]
@niklas-arm I added us ticker by default for mbed SDK  targets. As it is now, worked for my mbed SDK use. I haven't tested this branch wtih yotta tests.

**Comment by niklas-arm** [comment]
I just tested this and it works with yotta.

**Comment by PrzemekWirkus** [comment]
> I just tested this and it works with yotta.

@niklas-arm How about it works with us_ticker ? Did you try this solution.

**Comment by niklas-arm** [comment]
`us_ticker.h` was moved to `us_ticker_api.h`, that's why it broke, see https://github.com/ARMmbed/utest/commit/e39d77c38d25e50b4407643725b260cf2553af9d.

**Comment by 0xc0170** [comment]
> us_ticker.h was moved to us_ticker_api.h, that's why it broke, see e39d77c.

it was always ``us_ticker_api.h`` ?https://github.com/ARMmbed/mbed-hal/blob/master/mbed-hal/us_ticker_api.h

**Comment by niklarm** [comment]
> it was always us_ticker_api.h ?https://github.com/ARMmbed/mbed-hal/blob/master/mbed-hal/us_ticker_api.h

Huh? Did this ever compile? Nope. Classic case of "wasn't tested".
Fixed now.

**Pull request merged**